ts/src/connections/tunnelRelayTunnelHost.ts

```diff
@@ -224,19 +218,42 @@ export class TunnelRelayTunnelHost extends TunnelHostBase {
     }
 
     private onSshClientAuthenticated(session: SshServerSession) {
+        this.startForwardingExistingPorts(session);
+    }
+
+    private onSshClientReconnected(session: SshServerSession) {
+        this.startForwardingExistingPorts(session, true);
+    }
+
+    private async startForwardingExistingPorts(
+        session: SshServerSession,
+        removeUnusedPorts?: boolean,
+    ) {
         let pfs = session.activateService(PortForwardingService);
-        if (this.tunnel && this.tunnel.ports) {
-            this.tunnel.ports.forEach(async (port) => {
-                try {
-                    await this.forwardPort(pfs, port);
-                } catch (ex) {
-                    this.trace(
-                        TraceLevel.Error,
-                        0,
-                        `Error forwarding port ${port.portNumber}: ${ex}`,
-                    );
+        let ports = this.tunnel?.ports ?? [];
+        ports.forEach(async (port) => {
+            try {
+                await this.forwardPort(pfs, port);
+            } catch (ex) {
+                this.traceError(`Error forwarding port ${port.portNumber}: ${ex}`);
+            }
+        });
+
+        // If a tunnel client reconnects, its SSH session Port Forwarding service may
+        // have remote port forwarders for the ports no longer forwarded.
```

Comment:
Is this necessary? If host ports were refreshed, any remote forwarders for removed ports would have been dropped and disposed, and that would send a message to the client to stop forwarding the port. The message would be queued while the client is in a disconnected state, and sent after reconnection.

Reply:
If SSH session reconnection fails and after that regular connection succeeds, the queued messages won't be resent.

Reply:
After a non-SSH reconnection, any client sessions will be all new and will not have any port-forwarding state from before. None of the existing remote forwarder instances here will match the new session.